docs: refactor charm test how-tos (#3)
IronCore864/operator: docs/howto/write-unit-tests-for-a-charm.md

(write-unit-tests-for-a-charm)=
# How to write unit tests for a charm

> ## Setting up

Reviewer: How about:
```suggestion
## Setting up your environment
```

> First of all, install the Ops testing framework. To do this in a virtual environment while we are developing, use `pip` or other package managers. For example:

Reviewer: A couple of small suggestions:
```suggestion
First of all, install the Ops testing framework. To do this in a virtual environment while we're developing, use `pip` or a different package manager. For example:
```

```
pip install ops[testing]
```

> Normally, we'll include this and pin-point it to the latest minor version in the dependency group for our unit tests, for example in a `test-requirements.txt` file:

Reviewer: I think I get what you mean by "normally" here. To make it more explicit, how about this:
```suggestion
When we want to run repeatable unit tests, we'll normally pin `ops[testing]` to the latest minor version in the dependency group for our unit tests. For example in a `test-requirements.txt` file:
```

```text
ops[testing] ~= 2.19
```

Or in `pyproject.toml`:

```toml
[dependency-groups]
test = [
  "ops[testing] ~= 2.19",
]
```

## Creating the charm and test files

> Declare our new charm type in `charm.py`:

Reviewer: I think it would be good to add a little bit more context. How about:

```suggestion
So that we have a charm to test, declare a placeholder charm type in `charm.py`:
```

```python
class MyCharm(ops.CharmBase):
    pass        
```

> Open a new `test_foo.py` file where we will put the test code and import the [`ops.testing`](ops_testing) framework:

Reviewer: Minor suggestions for readability:

```suggestion
Then open a new `test_foo.py` file for the test code and import the [`ops.testing`](ops_testing) framework:
```

```python
import ops
from ops import testing
```

## Writing a test

To write a test function, use a `Context` object to encapsulate the charm type (`MyCharm`) and any necessary metadata. The test should then define the initial `State` and call `Context.run` with an `event` and initial `State`.

This follows the typical test structure:

> - Arrange: arrange inputs, mock necessary functions/system calls, initialize the charm
> - Act: act by calling `Context.run`
> - Assert: assert expected outputs or function calls.

Reviewer: I think this would be fine without the colons:

```suggestion
- Arrange inputs, mock necessary functions/system calls, and initialize the charm
- Act by calling `Context.run`
- Assert expected outputs or function calls.
```

> For example, if `MyCharm` writes a YAML config file via `Container.Push` on the pebble-ready event:

Reviewer: Suggest restructuring to avoid "via":

```suggestion
For example, suppose that `MyCharm` uses `Container.Push` to write a YAML config file on the pebble-ready event:
```

```python
def _on_pebble_ready(self, event: ops.PebbleReadyEvent):        
    container = event.workload
    container.push('/etc/config.yaml', 'message: Hello, world!', make_dirs=True)
    # ...
```

> And we want to test this behaviour, the test might look like this:

Reviewer: Suggest restructuring to match the previous sentence. Also switching to US English to match "initialize" earlier ijn the page:

```suggestion
A test for this behavior might look like:
```

```python
import yaml
from ops import testing

from charm import MyCharm

def test_pebble_ready_writes_config_file():
    """Test that on pebble-ready, a config file is written."""
    # Arrange: setting up the inputs
    ctx = testing.Context(MyCharm)
    container = testing.Container(name="some-container", can_connect=True)
    state_in = testing.State(
        containers=[container],
        leader=True,
    )

    # Act:
    ctx.run(ctx.on.pebble_ready(container=container), state_in)

    # Assert:

>     container_root_fs = container.get_filesystem(ctx)

Reviewer: Recently I learned that, although `container` works, when doing assertions it's better to get the container from `state_out.get_container("some-container")`.

Author: Thanks for pointing this out. As Tony mentioned to me earlier today, this might be a bug, we should definitely get containers from `state_out`.

    cfg_file = container_root_fs / "etc" / "config.yaml"
    config = yaml.safe_load(cfg_file.read_text())
    assert config["message"] == "Hello, world!"

```

```{note}

> If you like using unittest, you should rewrite this as a method of some TestCase subclass.

Reviewer: Minor suggestions:

```suggestion
If you prefer to use unittest, you should rewrite this as a method of a `TestCase` subclass.
```

```

> See more: 
>  - [`State`](ops.testing.State)
>  - [`Context`](ops.testing.Context)

## Mocking beyond the State

> If you wish to use the framework to test an existing charm type, you will probably need to mock out certain calls that are not covered by the `State` data structure. In that case, you will have to manually mock, patch or otherwise simulate those calls on top of what the framework does for you.

Reviewer: Both sentences kind of say the same thing, so I think we can shorten it:

```suggestion
If you wish to use the framework to test an existing charm type, you will probably need to mock out certain calls that are not covered by the `State` data structure. In that case, you will have to manually mock, patch or otherwise simulate those calls.
```

> For example, suppose that the charm we're testing uses the `KubernetesServicePatch`. To update the test above to mock that object, modify the test file to contain:

Reviewer: In the example, would it be possible to use something other than `KubernetesServicePatch`? I find the name distracting because it has "patch" in the name, but is nothing to do with patching in the context of testing!

Also, I think the example needs to have `from charm import MyCharm` to make sure that `MyCharm` can be yielded, doesn't it?

Author: Thanks for pointing it out, let me double-check this.
